libgdf/src/gdf_type_dispatcher.cuh

```diff
@@ -0,0 +1,60 @@
+#ifndef GDF_TYPE_DISPATCHER_H
+#define GDF_TYPE_DISPATCHER_H
+
+#include <utility>
+#include <gdf/cffi/types.h>
+
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief  Invokes a templated functor by dispatching the appropriate type from 
+ * a gdf_dtype enum value.
+ *
+ * This helper function accepts any *functor* (generic lambdas will also work, but
+ * only in host code.) with an "operator()" template. The template may have 1 or more
+ * template parameters, but the first parameter must be the type dispatched from the 
+ * gdf_dtype enum. The remaining template arguments must be able to be automatically 
+ * deduced. 
+ * 
+ * @Param dtype The gdf_dtype enum that determines which type will be dispatched
+ * @Param f The functor with a templated "operator()" that will be invoked with 
+ * the dispatched type
+ * @Param args A parameter-pack (i.e., arbitrary number of arguments) that will 
+ * be perfectly-forwarded as the arguments of the functor's "operator()".
+ *
+ * @Returns Whatever is returned by the functor's "operator()". 
+ *
+ * NOTE: The return type for all template instantiations of the functor's "operator()" 
+ * must be the same, otherwise there will be a compiler error as you would be 
+ * trying to return different types from the same function.
+ */
+/* ----------------------------------------------------------------------------*/
+// This pragma disables a compiler warning that complains about the valid usage
+// of calling a __host__ functor from this function which is __host__ __device__
+#pragma hd_warning_disable
+template <class functor_t, typename... Ts>
+__host__ __device__ __forceinline__
+decltype(auto) gdf_type_dispatcher(gdf_dtype dtype, functor_t f, Ts&&... args)
+{
+    switch(dtype)
+    {
+      // The .template is known as a "template disambiguator" 
+      // See here for more information: https://stackoverflow.com/questions/3786360/confusing-template-error
+      case GDF_INT8:      { return f.template operator()<int8_t>(std::forward<Ts>(args)...); }
```

Comment:
Does this work if the functor returns void?

Reply:
Hmmm, it works except for the default return.